src/Model.cpp

```diff
@@ -24,39 +29,49 @@ bool Edge::operator==(const Edge &other) const {
 }
 
 Model::Model()
-    : m_vnId(1) {
+    : d_ptr(new ModelPrivate())
+{
 }
 
-Model::~Model() {
+Model::Model(const Model &other)
+    : d_ptr(other.d_ptr)
+{
 }
 
-void Model::addChain(QSharedPointer<Chain> chain) {
-    if (!m_chains.contains(chain)) {
-        m_chains.append(chain);
-        emit chainsChanged(m_chains);
+Model::Model(QSharedPointer<ModelPrivate> other_ptr)
+    : d_ptr(other_ptr)
+{
+}
+
+void Model::addChain(Chain chain) {
+    if (!d_ptr->m_chains.contains(chain)) {
+        d_ptr->m_chains.append(chain);
+        emit chainsChanged(d_ptr->m_chains);
     }
 }
 
-void Model::removeChain(QSharedPointer<Chain> chain) {
-    if (m_chains.contains(chain)) {
-        m_chains.removeAll(chain);
-        emit chainsChanged(m_chains);
+void Model::removeChain(Chain chain) {
+    if (d_ptr->m_chains.contains(chain)) {
+        d_ptr->m_chains.removeAll(chain);
+        emit chainsChanged(d_ptr->m_chains);
     }
 }
 
-QList<QSharedPointer<Chain>> Model::chains() {
-    return m_chains;
+QList<Chain> Model::chains() {
+    return d_ptr->m_chains;
 }
 
 void Model::prepareNode(VideoNode *videoNode) {
     if(!videoNode)
         return;
 
+    // Giving a QSP a parent... hmm...
```

Comment:
It's weird because they are two different ways Qt has for managing object lifetimes, and you don't really mix them. Nothing technically wrong with this, just feels bad